File: src/core/Shaders/clearcoat.fs
// ===== CLEARCOAT BRDF (Conditional Compilation) =====
#ifdef ENABLE_CLEARCOAT

// Note: evaluateLayeredBRDF and calculateLayerAttenuation functions
// have been moved to material_evaluation.fs

// Improved clearcoat sampling function
vec3 sampleClearcoat( inout Ray ray, HitInfo hitInfo, RayTracingMaterial material, vec2 randomSample, out vec3 L, out float pdf, inout uint rngState ) {
	vec3 N = hitInfo.normal;
	vec3 V = - ray.direction;

    // Clamp clearcoat roughness to avoid artifacts
	float clearcoatRoughness = max( material.clearcoatRoughness, 0.089 );
	float baseRoughness = max( material.roughness, 0.089 );

    // Calculate sampling weights based on material properties
	float specularWeight = ( 1.0 - baseRoughness ) * ( 0.5 + 0.5 * material.metalness );
	float clearcoatWeight = material.clearcoat * ( 1.0 - clearcoatRoughness );
	float diffuseWeight = ( 1.0 - specularWeight ) * ( 1.0 - material.metalness );

    // Normalize weights
	float total = specularWeight + clearcoatWeight + diffuseWeight;
	specularWeight /= total;
	clearcoatWeight /= total;
	diffuseWeight /= total;

    // Choose which layer to sample
	float rand = RandomValue( rngState );
	vec3 H;

	if( rand < clearcoatWeight ) {
        // Sample clearcoat layer
		H = ImportanceSampleGGX( N, clearcoatRoughness, randomSample );
		L = reflect( - V, H );
	} else if( rand < clearcoatWeight + specularWeight ) {
        // Sample base specular
		H = ImportanceSampleGGX( N, baseRoughness, randomSample );
		L = reflect( - V, H );
	} else {
        // Sample diffuse
		L = ImportanceSampleCosine( N, randomSample );
		H = normalize( V + L );
	}

	// Calculate dot products
	DotProducts dots = computeDotProducts( N, V, L );

    // Calculate individual PDFs
	float clearcoatPDF = DistributionGGX( dots.NoH, clearcoatRoughness ) * dots.NoH / ( 4.0 * dots.VoH ) * clearcoatWeight;
	float specularPDF = DistributionGGX( dots.NoH, baseRoughness ) * dots.NoH / ( 4.0 * dots.VoH ) * specularWeight;
	float diffusePDF = dots.NoL / PI * diffuseWeight;

    // Combined PDF using MIS
	pdf = clearcoatPDF + specularPDF + diffusePDF;
	pdf = max( pdf, 0.001 ); // Ensure PDF is never zero

    // Evaluate complete BRDF
	return evaluateLayeredBRDF( dots, material );
}

#endif // ENABLE_CLEARCOAT

File: src/core/WebGPU/TSL/Clearcoat.js
// Clearcoat BRDF - Ported from clearcoat.fs
// Note: evaluateLayeredBRDF and calculateLayerAttenuation functions
// are in MaterialEvaluation.js

import {
	Fn,
	vec3,
	float,
	dot,
	normalize,
	reflect,
	max,
	If,
} from 'three/tsl';

import { struct } from './structProxy.js';

import { Ray, HitInfo, RayTracingMaterial, DotProducts } from './Struct.js';
import { PI, MIN_CLEARCOAT_ROUGHNESS, computeDotProducts } from './Common.js';
import { DistributionGGX } from './MaterialProperties.js';
import { ImportanceSampleGGX, ImportanceSampleCosine } from './MaterialSampling.js';
import { evaluateLayeredBRDF } from './MaterialEvaluation.js';
import { RandomValue } from './Random.js';

export const ClearcoatResult = struct( {
	brdf: 'vec3',
	L: 'vec3',
	pdf: 'float',
} );

// Improved clearcoat sampling function
// Returns vec4: xyz = brdf color, w = pdf
// L (light direction) is returned via the out pattern as a separate return
export const sampleClearcoat = Fn( ( [
	ray, hitInfo, material, randomSample, rngState,
] ) => {

	const N = hitInfo.normal;
	const V = ray.direction.negate();

	// Clamp clearcoat roughness to avoid artifacts
	const clearcoatRoughness = max( material.clearcoatRoughness, MIN_CLEARCOAT_ROUGHNESS );
	const baseRoughness = max( material.roughness, MIN_CLEARCOAT_ROUGHNESS );

	// Calculate sampling weights based on material properties
	const specularWeight = float( 1.0 ).sub( baseRoughness ).mul( float( 0.5 ).add( float( 0.5 ).mul( material.metalness ) ) ).toVar();
	const clearcoatWeight = material.clearcoat.mul( float( 1.0 ).sub( clearcoatRoughness ) ).toVar();
	const diffuseWeight = float( 1.0 ).sub( specularWeight ).mul( float( 1.0 ).sub( material.metalness ) ).toVar();

	// Normalize weights
	const total = specularWeight.add( clearcoatWeight ).add( diffuseWeight );
	specularWeight.divAssign( total );
	clearcoatWeight.divAssign( total );
	diffuseWeight.divAssign( total );

	// Choose which layer to sample
	const rand = RandomValue( rngState );

	const L = vec3( 0.0 ).toVar();
	const H = vec3( 0.0 ).toVar();

	If( rand.lessThan( clearcoatWeight ), () => {

		// Sample clearcoat layer
		H.assign( ImportanceSampleGGX( N, clearcoatRoughness, randomSample ) );
		L.assign( reflect( V.negate(), H ) );

	} ).ElseIf( rand.lessThan( clearcoatWeight.add( specularWeight ) ), () => {

		// Sample base specular
		H.assign( ImportanceSampleGGX( N, baseRoughness, randomSample ) );
		L.assign( reflect( V.negate(), H ) );

	} ).Else( () => {

		// Sample diffuse
		L.assign( ImportanceSampleCosine( N, randomSample ) );
		H.assign( normalize( V.add( L ) ) );

	} );

	// Calculate dot products
	const dots = DotProducts.wrap( computeDotProducts( N, V, L ) );

	// Calculate individual PDFs
	const clearcoatPDF = DistributionGGX( dots.NoH, clearcoatRoughness ).mul( dots.NoH ).div( float( 4.0 ).mul( dots.VoH ) ).mul( clearcoatWeight );
	const specularPDF = DistributionGGX( dots.NoH, baseRoughness ).mul( dots.NoH ).div( float( 4.0 ).mul( dots.VoH ) ).mul( specularWeight );
	const diffusePDF = dots.NoL.div( PI ).mul( diffuseWeight );

	// Combined PDF using MIS
	const pdf = max( clearcoatPDF.add( specularPDF ).add( diffusePDF ), 0.001 );

	// Evaluate complete BRDF
	const brdf = evaluateLayeredBRDF( dots, material );

	// Return brdf, L direction, and pdf packed together
	// Caller needs L and pdf - return as struct-like output
	// We pack: result.xyz = brdf, result.w = pdf, L stored in separate output
	return ClearcoatResult( { brdf, L, pdf } );

} );
